### PR TITLE
Trif 807

### PR DESCRIPTION
**Pull request by seanmcilvenna**

### CODE REVIEW

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js [hunk 1]
```diff
@@ -24,6 +24,113 @@ var templateEditViewModel = function (templateId, defaults) {
     self.AvailableExtensions = ko.observableArray([]);
     self.SelectedAvailableExtensionId = ko.observable();
 
+    self.duplicateNode = function () {
```
Comment (seanmcilvenna) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Recommend changing to "isDuplicateNode()"

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js [hunk 2]
```diff
@@ -24,6 +24,113 @@ var templateEditViewModel = function (templateId, defaults) {
...
+        else return false;
+    }
+
+    self.moveUp = function () {
+        if (!self.CurrentNode() || !self.CurrentNode().Constraint() || !self.Constraints()) {
+            return false;
+        }
+
+        var constraint = self.CurrentNode().Constraint();
+
```
Comment (seanmcilvenna) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Remove extra white-space here, please. I know.. I'm picky.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js [hunk 3]
```diff
@@ -24,6 +24,113 @@ var templateEditViewModel = function (templateId, defaults) {
...
+            for (var x = 0; x < siblings.length; x++) {
+                if (constraint.Context() === siblings[x].Context() && constraint.Id() !== siblings[x].Id()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    self.showMoveUp = function () {
```
Comment (seanmcilvenna) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: As good practice, we should have JSDoc comments on each of these public methods.
I have not been very good at it, either. In these cases, they will be simple... Just a description, no parameters. See http://usejsdoc.org/ for syntax on JSDoc if needed.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js [hunk 4]
```diff
@@ -24,6 +24,113 @@ var templateEditViewModel = function (templateId, defaults) {
...
+        if (!self.CurrentNode() || !self.CurrentNode().Constraint() || !self.Constraints()) {
+            return false;
+        }
+
+        var constraint = self.CurrentNode().Constraint();
+
+        var siblings = constraint.Parent() ? constraint.Parent().Children : self.Constraints;
+        var index = siblings.indexOf(constraint);
+
+        if (!index && index != 0) return false;
```
Comment (seanmcilvenna) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: is "index != 0" needed here?

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js [hunk 5]
```diff
@@ -24,6 +24,113 @@ var templateEditViewModel = function (templateId, defaults) {
...
+
+        if (!constraint.IsPrimitive()) {
+            for (var x = 0; x < self.Constraints().length; x++) {
+                if (constraint.Context() === self.Constraints()[x].Context() && constraint.Id() !== self.Constraints()[x].Id()) {
+                    return true;
+                }
+            }
+        } else {
+            var siblings = constraint.Parent().Children();
+            for (var x = 0; x < siblings.length; x++) {
```
Comment (seanmcilvenna) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Not required... but, you could consider using _ (underscore.js) here. I have recently added it to the library of scripts used by Trifolia. I could see _.first() being used here to determine if it should return true.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js (hunk 4 above)
Comment (ajagann1) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Yeah... On second look this is actually just flat wrong. Should be fixed now.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js (hunk 2 above)
Comment (ajagann1) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Done

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js (hunk 3 above)
Comment (ajagann1) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Added.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js (hunk 5 above)
Comment (ajagann1) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: I think you meant _.find() here? I made the change and committed it to the branch.

File: Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js (hunk 1 above)
Comment (ajagann1) [inline Trifolia.Web/Scripts/TemplateEdit/TemplateEditViewModel.js]: Done.

**Pull request closed without merging**